templates/single-quiz/question_type-gap-fill.php

```diff
@@ -26,7 +26,7 @@
 
     <span class="gapfill-answer-pre">
 
-        <?php echo apply_filters( 'sensei_answer_text', esc_html( $question_data[ 'gapfill_pre' ] ) ); ?>
+        <?php echo wp_kses_post( apply_filters( 'sensei_answer_text', $question_data[ 'gapfill_pre' ] ) ); ?>
```

Comment:
Could the removal of `esc_html` have backwards compatibility concerns? E.g. if the gapfill question has HTML code that is meant to be escaped and displayed.

Reply:
> E.g. if the gapfill question has HTML code that is meant to be escaped and displayed.

I can't fathom why someone would add HTML to the answer but then not want to display it. Adding HTML implies that they want that HTML to be visible. But if they don't, then yes, this will break. I'm OK with this if you are. 🙂 Especially given that this makes quizzes a bit more flexible.

Reply:
> I can't fathom why someone would add HTML to the answer but then not want to display it.

I agree, and this is what I was trying to say (but maybe we disagree on what "displaying" the HTML means). I'm talking about displaying the literal HTML code. But maybe that doesn't make much sense?

For example, if I add "This is a \<strong> tag" as part of my answer text, then I want users to see exactly that (i.e. I want the HTML to be escaped). I do not want them to see "This is a **tag**". I believe that removing the call to `esc_html` will cause the latter to be displayed instead of the former.

But maybe that's just because I'm a developer, and the rest of the world would want to actually display the bolded text rather than the HTML code...

Reply:
I see what you're saying. The problem is that if we change it to `esc_html`, then we are seriously restricting any third party code that could be hooking into the filter. Remember that it wasn't being escaped at all before.

So...which do you think is the lesser of two evils here?

Reply:
I wouldn't want to `esc_html` on the whole thing. I agree with your choice of `wp_kses_post` there. I was just thinking that the removal of `esc_html` around `$question_data[ 'gapfill_pre' ]` was a candidate for potential backwards compatibility problems.

So I personally think it would make the most sense to leave the `esc_html` where it was, and simply _add_ the call to `wp_kses_post`. That way if I user types in HTML code, that's what will be displayed (as it was before). But still, it won't affect the filter. We escape what the user typed into the form, and then allow the filters to do whatever they want with it, including adding actual HTML code which we do not escape (but we do run through `wp_kses_post`).

```php
echo wp_kses_post( apply_filters( 'sensei_answer_text', esc_html( $question_data[ 'gapfill_pre' ] ) ) );
```

That being said, if we remove the escaping, and a user does want to add literal HTML code, they could simply type in the escape characters themselves (e.g. `&lt;`). So at least we would be able to give them a workaround. That said, I wouldn't consider this to be the expected behaviour.

Reply:
Gotcha. Fixed in https://github.com/Automattic/sensei/pull/2296/commits/f63c36edc36b683df849bb628b33eafedf4da6ed.